Reorganize the guide

mark-i-m opened the pull request:
cc @rust-lang/wg-learning 

Feedback appreciated :)

spastorino commented:
@mark-i-m in my opinion this looks much better and makes way more sense :).

JohnTitor commented on src/SUMMARY.md:
```diff
@@ -45,15 +46,24 @@
         - [Lowering AST to HIR](./lowering.md)
         - [Debugging](./hir-debugging.md)
+    - [The MIR (Mid-level IR)](./mir/index.md)
+        - [HAIR MIR construction](./mir/construction.md)
```
```suggestion
        - [HAIR and MIR construction](./mir/construction.md)
```

Should be?

chrissimpkins commented on src/about-this-guide.md:
```diff
@@ -2,14 +2,19 @@
-end with useful reference information.
+development.
+
+There are six parts to this guide:
```
Maybe links to the start of these sections in this ToC list?

chrissimpkins commented on src/ast-validation.md:
```diff
@@ -0,0 +1,6 @@
+AST validation is the process of checking various correctness properties about
+the AST after macro expansion.
+
+**TODO**: write this chapter.
```
Open new IR's for each of these TODO's?

chrissimpkins commented on src/part-2-intro.md:
```diff
@@ -1,12 +1,15 @@
+end-to-end process of compilation _and_ readers interested in learning about a
+specific system they wish to contribute to. If anything is unclear, feel free
+to file an issue on the [rustc-dev-guide
+repo](https://github.com/rust-lang/rustc-dev-guide) or contact the compiler
```
Maybe take them directly to the issue tracker in this URL?
```suggestion
repo](https://github.com/rust-lang/rustc-dev-guide/issues) or contact the compiler
```

chrissimpkins commented on src/pat-exhaustive-checking.md:
```diff
@@ -0,0 +1,7 @@
+compiler will check that bindings are irrefutable when made and that match arms
+are exhaustive.
+
+**TODO**: write this chapter.
```
Add IR for this TODO

chrissimpkins commented:
I really like the new intro paragraphs.  +1!

tshepang commented on src/SUMMARY.md:
```diff
@@ -99,20 +104,24 @@
+    - [Parameter Environments](./param_env.md)
+
+- [Part 5: The Compiler Backend](./part-5-intro.md)
+    - [The MIR (Mid-level IR)](./mir/index.md)
```
This also appears in frontend. It also raises question... are operations on MIR, such as optimizations and conversion to llvm-ir considered frontend or backend.

Centril replied:
"middle end"?

tshepang replied:
@Centril is that a normal compiler term? Also makes me wonder if we should bother with [front,mid,back]-end terms.

mark-i-m replied:
Honestly, I don't think any of these is a very well defined term.

Backing off of this terminology in favor of something else might be a good idea. eg, we could do "Intermediate Represtations" and "Producing a Binary" as the part titles instead

Centril replied:
@tshepang "Frontend" and "Backend" are common terms in compilers, but "middle-end" is probably not. Still, it seems fairly intuitive in context?

Centril replied:
The main problem with the "frontend" and "backend" thing as I see it is that the line between them is fairly arbitrary. One could reasonably say (we don't) that LLVM bitcode generation is still the frontend, and the real backend actually starts inside LLVM.

tshepang replied:
I vote for getting rid of the terms... I don't find their usage useful. Perhaps, if we wanted to split things, we could use more meaningful terms:
```
- parsing
- analysis (includes typeck and borrowck)
- codegen
```
That fees better than
```
- front
- middle
- back
```

chrissimpkins replied:
> The main problem with the "frontend" and "backend" thing as I see it is that the line between them is fairly arbitrary. One could reasonably say (we don't) that LLVM bitcode generation is still the frontend, and the real backend actually starts inside LLVM.

Hmm... IMO, the most intuitive set of terms are "representations".  You have a beginning representation of a program as Rust source text, things happen to it, and you end with a linked binary representation of the program.  There are intermediate representations along the way to get there.  Could we abandon the "frontend", "middle", "backend" terms + associated connotations, and just use "Early intermediate representations", "Late intermediate representations" if we feel that ordering of the representation in the ToC helps one to understand how the compile process works?

I also like the idea of a "You are Here" overview diagram that is used at the beginning of each chapter so that a reader can understand where they are in the full sequence of stages and how that stage relates to adjacent stages.  Maybe this eliminates the need for the order terminology altogether?

Centril replied:
@chrissimpkins I agree with the focus on IRs :) 👍

mark-i-m replied:
Ok so how about:
- Source code representations
- Analysis
- From MIR to binary

And I do like @chrissimpkins's idea about a "you are here" map

tshepang replied:
*early* and *late* have same problem with *front* and *back*... one has to choose some (semi-)arbitrary line, and decide what's early and what's late

mark-i-m replied:
Also, it occurs to me that this goes well with the classic definition of a compiler: a program that takes code in language A and outputs code in language B.

Centril commented on src/SUMMARY.md:
```diff
@@ -45,15 +46,24 @@
     - [Panic Implementation](./panic-implementation.md)
     - [Macro expansion](./macro-expansion.md)
     - [Name resolution](./name-resolution.md)
+    - [AST Validation](./ast-validation.md)
```
Feature gate checking could use a mention here (it's a pass sorta like AST validation).

Centril commented on src/SUMMARY.md:
```diff
@@ -45,15 +46,24 @@
     - [Memory Management in Rustc](./memory.md)
+
+- [Part 3: The Compiler Frontend](./part-3-intro.md)
     - [Lexing and Parsing](./the-parser.md)
     - [`#[test]` Implementation](./test-implementation.md)
     - [Panic Implementation](./panic-implementation.md)
     - [Macro expansion](./macro-expansion.md)
     - [Name resolution](./name-resolution.md)
+    - [AST Validation](./ast-validation.md)
```
Consider indenting this into "The Syntax and AST"

mark-i-m replied:
Do you mean this and all of the above?

Centril replied:
Yep

mark-i-m commented:
Ok, I have pushed the following changes:
- Renamed the part titles as discussed above, and updated the intros
- Move the `rustc_interface` and `rustc_driver` parts to the beginning of Part 3 -- **feedback appreciated on this**
- Renamed "High-level overview of the compiler source" to "The compiler source code"

mark-i-m commented:
CI is finally green :green_heart:

tshepang commented on src/part-3-intro.md:
```diff
@@ -1,7 +1,9 @@
-that the compiler can work with easily. These are called intermediate
-representations.
+This part of the guide looks at the various ways the compiler transforms and
+represents source code. This part describes the process of taking raw source
```
I think first sentence must be removed... the second one says pretty much the same thing, only better.

tshepang commented on src/part-4-intro.md:
```diff
@@ -1,7 +1,8 @@
+properties of the code and to inform later stages. Typically, this is what people
+mean when they talk about "Rust's type system". This includes the
 representation, inference, and checking of types, the trait system, and the
 borrow checker. These analyses do not happen as one big pass or set of
```
Is borrow checking considered part of the type system? That's surprising, given that type checking and borrow checking are separate functions.

mark-i-m replied:
Lifetimes are part of the type of a definition, so I guess it is part of the "type system".

tshepang commented on src/part-5-intro.md:
```diff
@@ -7,13 +7,14 @@ changes.
 It's often useful to think of compilers as being composed of a _frontend_ and a
 _backend_  (though in rustc, there's not a sharp line between frontend and
 backend). The _frontend_ is responsible for taking raw source code, checking it
```
We can still avoid *front* and *back* terms, and one benefit is we don't have to talk about "sharp lines".

mark-i-m replied:
I tried to de-emphasize it and disambiguate it from "codegen backend". Let me know what you think.

tshepang replied:
I like

mark-i-m commented:
Thanks all! I think this is ready for a final review?

chrissimpkins commented on src/SUMMARY.md:
```diff
@@ -100,20 +108,24 @@
         - [Two-phase-borrows](./borrow_check/two_phase_borrows.md)
+    - [Parameter Environments](./param_env.md)
+
+- [Part 5: From MIR to binaries](./part-5-intro.md)
```
👍

chrissimpkins commented on src/SUMMARY.md:
```diff
@@ -46,15 +45,29 @@
-    - [Macro expansion](./macro-expansion.md)
-    - [Name resolution](./name-resolution.md)
+
+- [Part 3: Source Code Representations](./part-3-intro.md)
```
👍

chrissimpkins commented on src/overview.md:
```diff
@@ -0,0 +1,3 @@
+# Overview of the Compiler
+
+See https://github.com/rust-lang/rustc-dev-guide/pull/633
```
```suggestion
Coming soon!  Work is in progress on this chapter.  See https://github.com/rust-lang/rustc-dev-guide/pull/633 for the source and the [project README](https://github.com/rust-lang/rustc-dev-guide) for local build instructions.
```

mark-i-m replied:
Good idea

chrissimpkins commented on src/part-3-intro.md:
```diff
@@ -0,0 +1,8 @@
+# Part 3: Source Code Representations
```
Like the intro!

chrissimpkins commented on src/syntax-intro.md:
```diff
@@ -0,0 +1,8 @@
+# Syntax and the AST
+
+Working directly with source code is very inconvenient and error-prone. Thus,
```
👍

mark-i-m commented:
CI is green again
